### PR TITLE
TP-937: Add statistics for current progress of instance id calculation

### PR DESCRIPTION
**Pull request by pativa**
Adds one MBean with statistics for each collection that instanceId calculation is used for.

* MBean always show how many partitions the collection is currently ready to be loaded for
* When the calculation job is started, the statistics will be reset and show the current calculation progress (total rows and current progress)
* The statistics for each collection are collected in separate JMX ObjectNames called e.g. `collection_spaceObject`. This is done because one mirror can contain and calculate multiple collections.

### CODE REVIEW

**Review by askoog — APPROVED** [review]

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+class PersistedInstanceIdStatistics implements PersistedInstanceIdStatisticsMBean {
+
+	private final Set<Integer> readyForNumberOfPartitions = new ConcurrentHashSet<>();
+	private LongAdder calculationProgress;
+	private Long totalToCalculate;
+
+	public void resetStatisticsForJobExecution(Set<Integer> calculatingForPartitions) {
+		calculationProgress = new LongAdder();
+		totalToCalculate = 0L;
+		readyForNumberOfPartitions.removeIf((numberOfPartitions -> !calculatingForPartitions.contains(numberOfPartitions)));
```
Comment (oskarwiksten) [inline ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java:32]: If `readyForNumberOfPartitions` is `[3, 4]` and `calculatingForPartitions` is `[4, 5]`, then this will result in `readyForNumberOfPartitions` becoming `[4]`, is this intended? Or should this be written as `readyForNumberOfPartitions.removeAll(calculatingForPartitions)`?

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		readyForNumberOfPartitions.add(numberOfPartitions);
+	}
+
+	@Override
+	public int[] getReadyForNumberOfPartitions() {
+		return readyForNumberOfPartitions.stream().mapToInt(x -> x).sorted().toArray();
+	}
+
+	@Override
+	public Long getCalculationProgress() {
```
Comment (oskarwiksten) [inline ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java:53]: What is the use-case of having this number compared to the logging being done in `PersistedInstanceIdCalculationService`? When does use either?
If it is for answering the question "is everything done?" then maybe it would be better to have a getter that returns `boolean` instead?

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java [hunk 3]
```diff
@@ -141,6 +162,9 @@ private void calculatePersistedInstanceId(String collectionName, Set<Integer> nu
 		);
 		DocumentCollection collection = spaceMirror.getDocumentDb().getCollection(collectionName);
 
+		PersistedInstanceIdStatistics statistics = jobStatistics.computeIfAbsent(collectionName, x -> new PersistedInstanceIdStatistics());
```
Comment (oskarwiksten) [inline ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java:165]: When does the map here not already contain a key for `collectionName`? Is it for the case where `calculatePersistedInstanceId` is manually called for an uninitialized collection? In that case, does that mean that such a collection would not be present as a jmx mbean?

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java (hunk 2 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java:53]: Yes, the point is mainly to have a way of answering the question if everything is done. I used progress so that you get an idea of how long the calculation process will be (and could be plotted if needed, but not sure if there is a use case for that). So yes, choosing a boolean instead would maybe be an option for this.

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java [hunk 4]
```diff
@@ -172,7 +196,10 @@ private void calculatePersistedInstanceId(String collectionName, Set<Integer> nu
 			log.info("Step 2/3\tWill delete no longer used fields [{}]", String.join(", ", noLongerNeededFields));
 		}
 
-		try (Stream<List<Document>> batches = StreamUtils.buffer(collection.findByQuery(createQuery(BATCH_SIZE, fieldNamesToCalculate, noLongerNeededFields)), BATCH_SIZE)) {
+		Query query = createQuery(BATCH_SIZE, fieldNamesToCalculate, noLongerNeededFields);
+		statistics.setTotalToCalculate(collection.countByQuery(query));
```
Comment (oskarwiksten) [inline ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java:200]: Is it possible that the collection is modified in between the execution of this query and the next, which would lead to `totalToCalculate` being slightly wrong? What consequence would that lead to?

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java (hunk 1 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java:32]: Yes, this was intended. In a state where `[3, 4]` is already calculated and a new calculation is done for `[4, 5]`, it will still be ready for loading `[4]`. The index for `[3]` will be dropped when the calculation is started and will not work anymore. `[5]` will not be ready until the calculation is completed. However, if the current amount of partitions is `[4]`, it should be no problem restarting the instance, and continuing the calculation for `[5]` partitions where it ended.

`readyForNumberOfPartitions` is intended to show the number of instances the collection at this moment is ready for being restarted for.

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java (hunk 3 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java:165]: Yes, your right that it would not be present as a jmx mbean. I don't think this should be a usual case. I mainly wrote this so that `statistics` wouldn't be set to `null`, requiring null-checks in the calculation job.

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java (hunk 4 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java:200]: Yes, of course you're right there 🤦‍♂️  Simply using a `boolean` to determine if the calculation is ready as you suggested might be much simpler.

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java [hunk 5]
```diff
@@ -81,6 +83,25 @@ public void setApplicationContext(@Nonnull ApplicationContext applicationContext
...
 	}
 
+	PersistedInstanceIdStatisticsMBean collectStatistics(MirroredObject<?> mirroredObject) {
+		return jobStatistics.computeIfAbsent(mirroredObject.getCollectionName(), x -> new PersistedInstanceIdStatistics());
+	}
+
+	// Initializes readyForNumberOfPartitions for collections that are already calculated
+	void initializeStatistics() {
+		Set<Integer> numberOfPartitionsToCalculate = getNumberOfPartitionsToCalculate();
+		jobStatistics.replaceAll((collectionName, statistics) -> {
```
Comment (oskarwiksten) [inline ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java:93]: This method `initializeStatistics` currently seems to rely on `collectStatistics` having been called before, so that `jobStatistics` is populated with the collection names. This class already seems to have the responsibility of determining "which collections should have an index?" (see `calculatePersistedInstanceId()`), so maybe this method could reuse that responsibility here, to reduce the requirement that some other method must be called before this `initialize`-method?

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java (hunk 1 above)
Comment (oskarwiksten) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java:32]: ok!

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java (hunk 5 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java:93]: Good suggestion! Removing the requirement of calling the methods in order here reduces the complexity a bit. 👍

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java (hunk 2 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java]: I updated this to simply use a `boolean` instead now. As you say, if the need is to check the current progression, the logging from this service can be used instead.

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java (hunk 4 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java]: As this is now changed to a `boolean`, this problem is avoided.

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.avanza.ymer;
+
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.openspaces.core.util.ConcurrentHashSet;
```
Comment (oskarwiksten) [inline ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java:21]: Just wanted to check whether it is intentional that a gs collection is used here instead of a collection type from the jdk?

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationServiceMBean.java [hunk 7]
```diff
@@ -21,4 +21,11 @@
 
 	void calculatePersistedInstanceId(String collectionName);
 
+	/**
+	 * Returns number of partitions that all collections enabled for persisted instance id are currently
+	 * ready for.
+	 * When only one collection is used, this is the same as {@link PersistedInstanceIdStatisticsMBean#getReadyForNumberOfPartitions()}.
+	 */
+	int[] getAllCollectionsReadyForNumberOfPartitions();
```
Comment (oskarwiksten) [inline ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationServiceMBean.java:29]: Maybe it's subjective, but the name here is hard to parse for me. Would it be confusing if this was also called `getReadyForNumberOfPartitions`, or something like "get number of partitions that the data is prepared for"?

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java (hunk 2 above)
Comment (oskarwiksten) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java]: 👍

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java (hunk 4 above)
Comment (oskarwiksten) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java]: 👍

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java (hunk 5 above)
Comment (oskarwiksten) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java]: Nice. Simpler code and easier to follow.

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java (hunk 3 above)
Comment (oskarwiksten) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java]: ok!

File: ymer/src/main/java/com/avanza/ymer/YmerSpaceSynchronizationEndpoint.java [hunk 8]
```diff
@@ -128,6 +129,12 @@ void registerExceptionHandlerMBean() {
 	void registerPersistedInstanceIdCalculationServiceMBean() {
 		String name = "se.avanzabank.space.mirror:type=PersistedInstanceIdCalculationService,name=persistedInstanceIdCalculationService";
 		registerMbean(persistedInstanceIdCalculationService, name);
+		spaceMirror.getMirroredDocuments().stream()
+				.filter(MirroredObject::persistInstanceId)
+				.forEach(mirroredObject -> {
+					String statisticsBean = "se.avanzabank.space.mirror:type=PersistedInstanceIdCalculationService,name=collection_" + mirroredObject.getCollectionName();
+					registerMbean(persistedInstanceIdCalculationService.collectStatistics(mirroredObject), statisticsBean);
```
[the remark this answers is not in the document]
Comment (oskarwiksten) [reply ymer/src/main/java/com/avanza/ymer/YmerSpaceSynchronizationEndpoint.java:136]: 👍

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationServiceMBean.java (hunk 7 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationServiceMBean.java]: I agree, I changed this name to `getNumberOfPartitionsThatDataIsPreparedFor` and for the collections it's called `getNumberOfPartitionsThatCollectionIsPreparedFor`.

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java (hunk 6 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java]: No, that was a mistake 😓  I switched to using a collection from jdk instead.

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java [hunk 9]
```diff
@@ -81,6 +84,29 @@ public void setApplicationContext(@Nonnull ApplicationContext applicationContext
...
+				log.warn("Unable to initialize persisted instance id statistics for collection {}", collectionName, e);
+			}
+		});
+	}
+
+	private PersistedInstanceIdStatistics getStatisticsForCollection(String collectionName) {
+		return jobStatistics.computeIfAbsent(collectionName, x -> new PersistedInstanceIdStatistics());
+	}
+
 	public boolean collectionNeedsCalculation(String collectionName) {
```
Comment (oskarwiksten) [inline ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java:110]: It's not part of the scope of this MR, but is the result of this method really correct when also using `nextNumberOfInstances`? `noneMatch(isIndexForAny..(set))` is hard to parse for my old-person-brain. 

Here is a test that exemplifies it, but I have probably misunderstood it:
```java
@Test
public void test() throws Exception {
	var mongoDbFactory = mirrorEnvironment.getMongoTemplate().getMongoDbFactory();
	var collectionName = TEST_SPACE_OBJECT.collectionName();
	mongoDbFactory.getDb().getCollection(collectionName)
			.createIndex(new Document(getInstanceIdFieldName(1), 1));

	var testSpaceMirrorFactory = new TestSpaceMirrorFactory(mongoDbFactory);
	testSpaceMirrorFactory.setNextNumberOfInstances(9);

	try (var endpoint = (YmerSpaceSynchronizationEndpoint) testSpaceMirrorFactory.createSpaceSynchronizationEndpoint()) {
		var target = endpoint.getPersistedInstanceIdCalculationService();
		execute(
				() -> assertTrue(target.collectionNeedsCalculation(collectionName)),
				new SystemProperties("cluster.partitions", "1")
		);
	}
}
```

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java (hunk 6 above)
Comment (oskarwiksten) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdStatistics.java]: 👍

File: ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java (hunk 9 above)
Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java:109]: No, you're right as your test exemplifies 😱  Thanks for finding this! This testcase also seems to be missing...
As you say, this is a part of an earlier PR (https://github.com/AvanzaBank/ymer/pull/66), but as that one is closed I'll fix it here.

This method was not working when multiple number of partitions were used. The result would have been that the job sometimes wouldn't have started correctly (if it already was calculated for the current number of partitions) 
 when a next number of instances is set.

I have pushed a fix for this bug now.

Comment (pativa) [reply ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdCalculationService.java:110]: I decided to remove the bugfix from this issue (as it's not really part of this PR), and open a separate PR: https://github.com/AvanzaBank/ymer/pull/69

So if you are done with the review of this issue, can you mark it as ready? 🙏

**Review by oskarwiksten — APPROVED** [review]
